### DISCUSSION TITLE
使用pg作为元数据引擎时，在已经格式化的情况下提示：database is not formatted，please run juicefs format

### DISCUSSION THREAD

**Discussion by ForgQi** [juicedata/juicefs#2636] [Q&A] [answered]
使用redis时正常，但使用pg时，format正常，mount时会提示未format，该如何排查问题在哪

**Comment by ForgQi** [comment]
你好，经过测试是schema导致的，查看源码xorm默认的schema是`public`
```go
func (db *postgres) getSchema() string {
	if db.uri.Schema != "" {
		return db.uri.Schema
	}
	return DefaultPostgresSchema
}
```
juicefs在format时使用了用户默认的schema，但随后load时,却使用xorm默认的`public`，导致可以format但是不能mount
```go
func (m *dbMeta) doLoad() (data []byte, err error) {
	err = m.roTxn(func(ses *xorm.Session) error {
		if ok, err := ses.IsTableExist(&setting{}); err != nil {
			return err
		} else if !ok {
			return nil
		}
		s := setting{Name: "format"}
		ok, err := ses.Get(&s)
		if err == nil && ok {
			data = []byte(s.Value)
		}
		return err
	})
	return
}
```
xorm:
```go
func (db *postgres) IsTableExist(queryer core.Queryer, ctx context.Context, tableName string) (bool, error) {
	if len(db.getSchema()) == 0 {
		return db.HasRecords(queryer, ctx, `SELECT tablename FROM pg_tables WHERE tablename = $1`, tableName)
	}

	return db.HasRecords(queryer, ctx, `SELECT tablename FROM pg_tables WHERE schemaname = $1 AND tablename = $2`,
		db.getSchema(), tableName)
}
```
希望可以修复这个问题，比如提供选项或者param设置schema，或者可以正确使用当前用户默认的schema，保证format与mount一致

**Comment by zhijian-pro** [comment]
能否提供下复现的命令参数？
pg 的 format mount 参数 都是按照这个文档所写的格式吗？https://juicefs.com/docs/zh/community/databases_for_metadata#postgresql

**Reply by ForgQi** [reply]
format mount和都是按照文档里所写，经过验证：使用`postgres://user:mypassword@xxx:5432/juicefs?search_path=public`去format和mount没有问题，结合源码基本可以确定是schema没有支持好，产生了不一致。

要复现可新建一个pg用户，查看或设置schema为`$user`，然后使用文档里的链接挂载`"postgres://user:mypassword@192.168.1.6:5432/juicefs"`即可:
```sql
ALTER USER username0 SET search_path TO '"$user", public';
SHOW search_path ;
```

**Comment by zhijian-pro** [comment, marked as answer]
的确会有这个问题，因为 xorm 对 search_path 没有支持。为了支持在非 public 的 schema 中使用 juicefs ，我们会对 search_path 进行限制，只允许用户填一个 schema。?search_path=juicefs,public 将不会被允许。限制后，juicefs 使用的 schema 默认为 public，如果用户设置了 search_path 则 schema 改为 search_path 中设置的 schema。
比如设置 ?search_path=pguser1 juicefs 则使用 pguser1这个 schema 进行操作

这个问题将会在 https://github.com/juicedata/juicefs/pull/2663 中解决
